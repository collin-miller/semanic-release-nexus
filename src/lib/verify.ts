import Ajv from 'ajv';
import AggregateError from 'aggregate-error';
import glob from 'tiny-glob';
import { Context } from 'semantic-release';
import { IPluginAssets, IPluginConfig, PluginSchema, resolveOptions } from './plugin';

export const isValidUrl = (url: string | undefined) => {
    try {
        if (url) {
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const _ = new URL(url);
        } else {
            return false;
        }
    } catch (e) {
        return false;
    }
    return true;
};

export const verifyAssets = (assets: Array<IPluginAssets>, context: Context): string[] => {
    const errors: string[] = [];
    assets.forEach(async (element) => {
        const files = await glob(element.path, { filesOnly: true });
        if (files.length === 0) {
            errors.push(`No asset was found at path ${element.path}`);
        } else if (files.length > 1) {
            errors.push(`Only one file supported per asset declaration. Files found ${files.join(' ')}`);
        } else {
            context.logger.log(`Found file ${files.join('')} from path ${element.path}`);
        }
    });
    return errors;
};

/**
 * A method to verify the inputs
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const verify = async (pluginConfig: IPluginConfig, context: Context) => {
    // const { logger } = context;
    const errors: string[] = [];

    const ajv = new Ajv({ allErrors: true });
    const valid = ajv.validate(PluginSchema, pluginConfig);
    // Collect schema validation errors
    if (!valid) {
        ajv.errors?.forEach((element) => errors.push(element.message || 'validation error'));
    }
    // verify a valid file is found at each asset path
    errors.push(...verifyAssets(pluginConfig.assets, context));
    const options = resolveOptions(pluginConfig, context);
    if (!isValidUrl(options.nexusHost)) {
        errors.push(`The nexus host provided (${options.nexusHost}) is invalid.`);
    }
    if (typeof options.nexusRepo === 'undefined') {
        errors.push(`The nexus repo (${options.nexusRepo}) was not provided.`);
    }
    if (errors.length > 0) {
        throw new AggregateError(errors);
    }
};
