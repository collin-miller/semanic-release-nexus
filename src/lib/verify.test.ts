import { afterEach, describe, expect, jest, it } from '@jest/globals';
import AggregateError from 'aggregate-error';
import glob from 'tiny-glob';
import { verify } from './verify';

jest.mock('tiny-glob');

describe('verify', () => {
    const mockedGlob = glob as jest.Mocked<typeof glob>;
    afterEach(() => {
        (mockedGlob as jest.Mock).mockReset();
    });
    it('should fail schema validation', async () => {
        await expect(
            verify({ assets: [] }, { logger: { log: jest.fn(), error: jest.fn() }, env: { key: 'value' } }),
        ).rejects.toThrowError(AggregateError);
    });
    it('should pass schema validation', async () => {
        (mockedGlob as jest.Mock)
            .mockImplementationOnce(() => ['dist/some-package-v1.tar.gz'])
            .mockImplementationOnce(() => ['dist/other-package-v2.tar.gz']);
        const logMock = jest.fn();
        await expect(
            verify(
                {
                    assets: [{ path: 'dist/some-package-v*.tar.gz' }, { path: 'dist/other-package-v*.tar.gz' }],
                },
                { logger: { log: logMock, error: jest.fn() }, env: { key: 'value' } },
            ),
        ).resolves.toBeUndefined();
        expect(mockedGlob).toHaveBeenCalledTimes(2);
        expect(mockedGlob).toHaveBeenNthCalledWith(1, 'dist/some-package-v*.tar.gz', { filesOnly: true });
        expect(mockedGlob).toHaveBeenNthCalledWith(2, 'dist/other-package-v*.tar.gz', { filesOnly: true });
        expect(logMock).toBeCalled();
        expect(logMock.mock.calls).toMatchSnapshot();
    });
    it('should fail schema validation for an additional asset map', async () => {
        (mockedGlob as jest.Mock)
            .mockImplementationOnce(() => ['dist2/some-package-v1.tar.gz'])
            .mockImplementationOnce(() => ['dist2/other-package-v2.tar.gz']);
        const logMock = jest.fn();
        await expect(
            verify(
                {
                    assets: [
                        { path: 'dist2/some-package-v*.tar.gz' },
                        // @ts-ignore: invalid type
                        { path: 'dist2/other-package-v*.tar.gz', booger: true },
                    ],
                },
                { logger: { log: logMock, error: jest.fn() }, env: { key: 'value' } },
            ),
        ).rejects.toThrowError(AggregateError);
        expect(mockedGlob).toHaveBeenCalledTimes(2);
        expect(mockedGlob).toHaveBeenNthCalledWith(1, 'dist2/some-package-v*.tar.gz', { filesOnly: true });
        expect(mockedGlob).toHaveBeenNthCalledWith(2, 'dist2/other-package-v*.tar.gz', { filesOnly: true });
        expect(logMock).toBeCalled();
        expect(logMock.mock.calls).toMatchSnapshot();
    });
});
