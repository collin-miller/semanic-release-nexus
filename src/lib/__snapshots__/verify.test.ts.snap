// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify should fail schema validation for an additional asset map 1`] = `
Array [
  Array [
    "Found file dist2/some-package-v1.tar.gz from path dist2/some-package-v*.tar.gz",
  ],
  Array [
    "Found file dist2/other-package-v2.tar.gz from path dist2/other-package-v*.tar.gz",
  ],
]
`;

exports[`verify should pass schema validation 1`] = `
Array [
  Array [
    "Found file dist/some-package-v1.tar.gz from path dist/some-package-v*.tar.gz",
  ],
  Array [
    "Found file dist/other-package-v2.tar.gz from path dist/other-package-v*.tar.gz",
  ],
]
`;
